File: app/src/main/java/com/elishaazaria/sayboard/preferences/ModelPreferences.java
package com.elishaazaria.sayboard.preferences;

import com.elishaazaria.sayboard.AppCtx;
import com.elishaazaria.sayboard.R;
import com.elishaazaria.sayboard.data.VoskServerData;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ModelPreferences {
    public static final boolean VOSK_SERVER_ENABLED = false;

    public static List<VoskServerData> getVoskServers() {
        Set<String> set = MyPreferences.getSharedPref().getStringSet(AppCtx.getStringRes(R.string.pref_models_vosk_servers_set),
                null);

        if (set == null) return new ArrayList<>();

        ArrayList<VoskServerData> list = new ArrayList<>();

        for (String stringData : set) {
            VoskServerData data = VoskServerData.deserialize(stringData);
            if (data != null) {
                list.add(data);
            }
        }

        Collections.sort(list);
        return list;
    }

    public static void setVoskServers(List<VoskServerData> servers) {
        Set<String> set = new HashSet<>();
        for (VoskServerData data : servers) {
            set.add(VoskServerData.serialize(data));
        }

        MyPreferences.getSharedPref().edit().putStringSet(AppCtx.getStringRes(R.string.pref_models_vosk_servers_set), set).apply();
    }

    public static void addToVoskServers(VoskServerData data) {
        Set<String> set = new HashSet<>(MyPreferences.getSharedPref().getStringSet(AppCtx.getStringRes(R.string.pref_models_vosk_servers_set),
                new HashSet<>()));
        set.add(VoskServerData.serialize(data));
        MyPreferences.getSharedPref().edit().putStringSet(AppCtx.getStringRes(R.string.pref_models_vosk_servers_set), set).apply();
    }

    public static void removeFromVoskServers(VoskServerData data) {
        Set<String> set = new HashSet<>(MyPreferences.getSharedPref().getStringSet(AppCtx.getStringRes(R.string.pref_models_vosk_servers_set),
                new HashSet<>()));
        set.remove(VoskServerData.serialize(data));
        MyPreferences.getSharedPref().edit().putStringSet(AppCtx.getStringRes(R.string.pref_models_vosk_servers_set), set).apply();
    }
}

File: app/src/main/java/com/elishaazaria/sayboard/preferences/ModelPreferences.kt
package com.elishaazaria.sayboard.preferences

import com.elishaazaria.sayboard.AppCtx.getStringRes
import com.elishaazaria.sayboard.AppCtx.getBoolRes
import com.elishaazaria.sayboard.AppCtx.appCtx
import com.elishaazaria.sayboard.AppCtx.getIntegerRes
import com.elishaazaria.sayboard.preferences.MyPreferences
import com.elishaazaria.sayboard.AppCtx
import com.elishaazaria.sayboard.R
import com.elishaazaria.sayboard.preferences.LogicPreferences
import com.elishaazaria.sayboard.data.VoskServerData
import android.content.SharedPreferences
import android.os.Build
import com.elishaazaria.sayboard.preferences.UIPreferences
import androidx.core.content.ContextCompat
import java.util.*
import kotlin.collections.HashSet

object ModelPreferences {
    const val VOSK_SERVER_ENABLED = false
    @JvmStatic
    var voskServers: List<VoskServerData>
        get() {
            val set = MyPreferences.sharedPref.getStringSet(
                getStringRes(R.string.pref_models_vosk_servers_set),
                null
            ) ?: return ArrayList()
            val list = ArrayList<VoskServerData>()
            for (stringData in set) {
                val data = VoskServerData.deserialize(stringData)
                if (data != null) {
                    list.add(data)
                }
            }
            Collections.sort(list)
            return list
        }
        set(servers) {
            val set: MutableSet<String> = HashSet()
            for (data in servers) {
                set.add(VoskServerData.serialize(data))
            }
            MyPreferences.sharedPref.edit()
                .putStringSet(getStringRes(R.string.pref_models_vosk_servers_set), set).apply()
        }

    fun addToVoskServers(data: VoskServerData) {
        val set: MutableSet<String> = HashSet(
            MyPreferences.sharedPref.getStringSet(
                getStringRes(R.string.pref_models_vosk_servers_set),
                HashSet()
            )
        )
        set.add(VoskServerData.serialize(data))
        MyPreferences.sharedPref.edit()
            .putStringSet(getStringRes(R.string.pref_models_vosk_servers_set), set).apply()
    }

    fun removeFromVoskServers(data: VoskServerData) {
        val set: MutableSet<String> = HashSet(
            MyPreferences.sharedPref.getStringSet(
                getStringRes(R.string.pref_models_vosk_servers_set),
                HashSet()
            )
        )
        set.remove(VoskServerData.serialize(data))
        MyPreferences.sharedPref.edit()
            .putStringSet(getStringRes(R.string.pref_models_vosk_servers_set), set).apply()
    }
}